File: playwright/pages/inventory.page.ts
import { Locator, Page } from '@playwright/test'

export class InventoryPage {
    readonly page: Page;
    readonly lista: Locator;
    readonly nomeDosProtudos: Locator;
    readonly contadorDoCarrinho: Locator;
    readonly linkDoCarrinho: Locator;
    readonly seletorDeOrdenacao: Locator;
    readonly menuHamburguer: Locator;
    readonly linkDeLogout: Locator;

    constructor(page: Page) {
        this.page = page;
        this.lista = page.locator('[data-test="inventory-list"]')
        this.nomeDosProtudos = page.locator('[data-test="inventory-item-name"]')
        this.contadorDoCarrinho = page.locator('[data-test="shopping-cart-badge"]')
        this.linkDoCarrinho = page.locator('[data-test="shopping-cart-link"]')
        this.seletorDeOrdenacao = page.locator('[data-test="product-sort-container"]')
        this.menuHamburguer = page.locator('#react-burger-menu-btn')
        this.linkDeLogout = page.locator('[data-test="logout-sidebar-link"]')
    }

    botaoDeAdicao(produto: string): Locator {
        return this.page.locator(`[data-test="add-to-cart-${produto}"]`)
    }

    botaoDeRemocao(produto: string): Locator {
        return this.page.locator(`[data-test="remove-${produto}"]`)
    }

    async acessarDiretamente(): Promise<void> {
        await this.page.goto('/inventory.html');
    }

    async adicionarAoCarrinho(produto: string): Promise<void> {
        await this.botaoDeAdicao(produto).click();
    }

    async removerDoCarrinho(produto: string): Promise<void> {
        await this.botaoDeRemocao(produto).click();
    }

    async abrirCarrinho(): Promise<void> {
        await this.linkDoCarrinho.click();
    }

    async ordenarPor(rotulo: string): Promise<void> {
        await this.seletorDeOrdenacao.selectOption({ label: rotulo });
    }

    async listarNomes(): Promise<string[]> {
        return this.nomeDosProtudos.allInnerTexts();
    }

    async sair(): Promise<void> {
        await this.menuHamburguer.click();
        await this.linkDeLogout.click();
    }
}

File: selenium/src/test/java/pages/InventoryPage.java
package pages;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

public class InventoryPage extends BasePage {

    private static final String URL = "https://www.saucedemo.com/inventory.html";

    private static final By LISTA_DE_PRODUTOS = By.cssSelector("[data-test='inventory-list']");
    private static final By NOMES_DOS_PRODUTOS = By.cssSelector("[data-test='inventory-item-name']");
    private static final By BADGE_DO_CARRINHO = By.cssSelector("[data-test='shopping-cart-badge']");
    private static final By LINK_DO_CARRINHO = By.cssSelector("[data-test='shopping-cart-link']");
    private static final By SELETOR_DE_ORDENACAO = By.cssSelector("[data-test='product-sort-container']");
    private static final By MENU_HAMBURGUER = By.id("react-burger-menu-btn");
    private static final By LINK_DE_LOGOUT = By.cssSelector("[data-test='logout-sidebar-link']");

    public InventoryPage(WebDriver driver) {
        super(driver);
    }

    private static By botaoDeAdicao(String produto) {
        return By.cssSelector("[data-test='remove-" + produto + "']");
    }

    private static By botaoDeRemocao(String produto) {
        return By.cssSelector("[data-test='add-to-cart-" + produto + "']");
    }

    public InventoryPage acessarDiretamente() {
        driver.get(URL);
        return this;
    }

    public boolean estaExibida() {
        return visivel(LISTA_DE_PRODUTOS).isDisplayed();
    }

    public InventoryPage adicionarAoCarrinho(String produto) {
        clicar(botaoDeAdicao(produto));
        return this;
    }

    public InventoryPage removerDoCarrinho(String produto) {
        clicar(botaoDeRemocao(produto));
        return this;
    }

    public String textoDoBotaoDeAdicao(String produto) {
        return texto(botaoDeAdicao(produto));
    }

    public String textoDoBotaoDeRemocao(String produto) {
        return texto(botaoDeRemocao(produto));
    }

    public String contadorDoCarrinho() {
        return texto(BADGE_DO_CARRINHO);
    }

    public boolean contadorDoCarrinhoAusente() {
        return estaAusente(BADGE_DO_CARRINHO);
    }

    public CartPage abrirCarrinho() {
        clicar(LINK_DO_CARRINHO);
        return new CartPage(driver);
    }

    public InventoryPage ordenarPor(String rotulo) {
        new Select(visivel(SELETOR_DE_ORDENACAO)).selectByVisibleText(rotulo);
        return this;
    }

    public List<String> nomesDosProdutos() {
        visivel(LISTA_DE_PRODUTOS);
        return driver.findElements(NOMES_DOS_PRODUTOS).stream()
                .map(WebElement::getText)
                .toList();
    }

    public LoginPage sair() {
        clicar(MENU_HAMBURGUER);
        clicar(LINK_DE_LOGOUT);
        return new LoginPage(driver);
    }
}
